### PR TITLE
Implement a few more rvalue translation to smir

### PR DESCRIPTION
**Pull request by celinval**
Add the implementation for a few more RValue variants. For now, I simplified the stable version of `RValue::Ref` by removing the notion of Region.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_smir/src/rustc_internal/mod.rs [hunk 1]
```diff
@@ -10,7 +10,7 @@ use crate::{
 use rustc_middle::ty::TyCtxt;
 pub use rustc_span::def_id::{CrateNum, DefId};
 
-fn with_tables<R>(mut f: impl FnMut(&mut Tables<'_>) -> R) -> R {
+pub(crate) fn with_tables<R>(mut f: impl FnMut(&mut Tables<'_>) -> R) -> R {
```
Comment (oli-obk) [inline compiler/rustc_smir/src/rustc_internal/mod.rs:13]: I wanted to keep this private for now, to have a stronger separation. Not sure yet whether that is good long term, but maybe let's keep it for now?

**Comment by oli-obk** [comment]
> by removing the notion of Region.

one thing I've done for cases like this is to just use a `String` as the "stable" version, so the data is at least there in some opaque, but printable, manner.

r=me after splitting off the Cargo.toml and extern crate changes into their own commit (and maybe the `Stable` cleanups that do not add new behaviour, too?)

@bors delegate+

**Comment by celinval** [comment]
> > by removing the notion of Region.
> 
> one thing I've done for cases like this is to just use a `String` as the "stable" version, so the data is at least there in some opaque, but printable, manner.

That makes sense. What if I create an actual Opaque type that implements ToString (or any other name you might prefer). Just to differentiate from cases where String is the actual stable version.

> r=me after splitting off the Cargo.toml and extern crate changes into their own commit (and maybe the `Stable` cleanups that do not add new behaviour, too?)

Ok. Just to clarify. Do you want me to split them into separate commit or separate PRs?

> @bors delegate+

File: compiler/rustc_smir/src/rustc_internal/mod.rs (hunk 1 above)
Comment (celinval) [reply compiler/rustc_smir/src/rustc_internal/mod.rs:13]: Ok, do you have any suggestion on how to get the context when translating from an internal rustc type that requires context information?

From what I can tell, we need to modify the Tables when translation from MIR -> SMIR and we need to read the Tables when translating SMIR -> MIR. So I was thinking that this function should be available for the `rustc_smir` and the `rustc_internal` to do their respective translations.

**Comment by oli-obk** [comment]
> What if I create an actual Opaque type that implements ToString (or any other name you might prefer).

Oh heh, that's even better, as then you have an actual type that can even be used in a `Stable` impls in a way that won't require changing anything but the type and the stable impls in the future when this gets turned into a real conversion.



> Do you want me to split them into separate commit or separate PRs?

Just separate commits. It's totally alright to piggyback some cleanup commits into a regular PR

File: compiler/rustc_smir/src/rustc_internal/mod.rs (hunk 1 above)
Comment (celinval) [reply compiler/rustc_smir/src/rustc_internal/mod.rs:13]: @oli-obk?

Comment (oli-obk) [reply compiler/rustc_smir/src/rustc_internal/mod.rs:13]: The wrapper function you removed was my attempt at keeping everything but the wrappers in that module. As long as that isn't too annoying, I think we should keep adding such wrapper functions as needed. We can always remove all of them and expose the tables function directly, but that should be its own PR with its own motivation.

Comment (celinval) [reply compiler/rustc_smir/src/rustc_internal/mod.rs:13]: That makes sense. Let me revert that. Thanks

**Comment by oli-obk** [comment]
bors silently ignores approvals that happen before (force-)pushes

@bors r+

**Comment by matthiaskrgr** [comment]
@bors r- 
might need rebase or something
https://github.com/rust-lang/rust/actions/runs/5358354083/jobs/9720522386#step:25:1244

**Comment by celinval** [comment]
@oli-obk I had to solve some conflicts. Do I need another approval?

**Comment by matthiaskrgr** [comment]
may need rebase

**Comment by celinval** [comment]
@oli-obk sorry, another conflict...

**Comment by oli-obk** [comment]
@bors r+ rollup

No other smir stuff in the queue, should work this time 😆

**Comment by oli-obk** [comment]
@bors rollup

It cannot possibly affect clippy. No one uses the SMIR crate.

**Pull request merged**